src/rpp/rpp/operators/take_until.hpp

```diff
@@ -0,0 +1,132 @@
+//                  ReactivePlusPlus library
+//
+//          Copyright Aleksey Loginov 2022 - present.
+//                    TC Wang 2022 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/defs.hpp>
+#include <rpp/operators/fwd/take_until.hpp>
+#include <rpp/operators/first.hpp>
+#include <rpp/sources/create.hpp>
+#include <rpp/subscribers/constraints.hpp>
+#include <rpp/utils/functors.hpp>
+
+
+#include <rpp/operators/details/subscriber_with_state.hpp> // create_subscriber_with_state
+
+
+IMPLEMENTATION_FILE(take_until_tag);
+
+namespace rpp::details
+{
+
+/**
+ * Functor (type-erasure) of "take_until" for on_next operator.
+ */
+template<constraint::decayed_type Type>
+struct take_until_on_next
+{
+    void operator()(auto&& value,
+                    const auto& subscriber,
+                    const std::shared_ptr<first_state>& state) const
+    {
+        if (state->count > 0)
+            subscriber.on_next(std::forward<decltype(value)>(value));
+        else
+            subscriber.on_completed();
+    }
+};
+
+template<constraint::decayed_type Type>
+struct take_until_throttler
+{
+
+    /**
+     * Functor (type-erasure) of "take_until" for on_next operator.
+     */
+    void operator()(auto&&,
+                    const auto& subscriber,
+                    const std::shared_ptr<first_state>& state) const
+    {
+        if (state->count > 0)
+            --state->count;
+
+        if (state->count == 0)
+            subscriber.on_completed();
+
+//        decrement_notification_count(subscriber, state);
+    }
+
+//    /**
+//     * Functor (type-erasure) of "take_until" for on_error operator.
+//     */
+//    void operator()(const std::exception_ptr&,
+//                    const auto& subscriber,
+//                    const std::shared_ptr<first_state>& state) const
+//    {
+//        decrement_notification_count(subscriber, state);
+//    }
+//
+//    /**
+//     * Functor (type-erasure) of "take_until" for on_completed operator.
+//     */
+//    void operator()(const auto& subscriber,
+//                    const std::shared_ptr<first_state>& state) const
+//    {
+//        decrement_notification_count(subscriber, state);
+//    }
+//
+//    void decrement_notification_count(const auto& subscriber,
+//                                      const std::shared_ptr<first_state>& state) const
+//    {
+//        if (state->count > 0)
+//            --state->count;
+//
+//        if (state->count == 0)
+//            subscriber.on_completed();
+//    }
```

Comment:
@victimsnino 

> The TakeUntil subscribes and begins mirroring the source Observable. It also monitors a second Observable that you provide. If this second Observable emits an item or sends a termination notification, the Observable returned by TakeUntil stops mirroring the source Observable and terminates.

I'm not sure if `take_until()` will translate an error event to a completed event and hence I have this commented code.

You could check out the unit test and if the unit test is correct, I'll remove these lines.

Reply:
Looks like your current implementation is correct (until observable should forward notification to subscriber)